glean-core/ffi/src/lib.rs

```diff
@@ -37,17 +37,28 @@ lazy_static! {
 type RawStringArray = *const *const c_char;
 
 /// Create a vector of strings from a raw C-like string array
-unsafe fn from_raw_string_array(arr: RawStringArray, len: i32) -> Vec<String> {
-    if arr.is_null() || len == 0 {
-        return vec![];
+///
+/// ## Safety
+///
+/// * We check the array pointer for validity (non-null)
+/// * FfiStr checks each individual char pointer for validity (non-null)
+/// * We discard invalid char pointers (null pointer)
+/// * invalid UTF-8 is replaced with the replacement character (guaranteed by ffi-support)
+fn from_raw_string_array(arr: RawStringArray, len: i32) -> Vec<String> {
+    unsafe {
```

Comment:
Curiosity: is it common practice to wrap the whole function in `unsafe` rather then restricting this to the "scary parts"? I could imagine `arr.is_null()` being kind of safe.

Reply:
Yes and no. Preference differs.
I do this because:

* `unsafe` marks a block where _some_ parts can't be proven by the compiler (it still does tons of checks though ;))
* `unsafe` does not introduce unsafe code, but merely saying "what follows the compiler can't proof safe, but I, the dev, guarantee that all invariants are held and that no _safe_ use of this block can cause unsafety"

The code _around_ the actual unsafe function calls (e.g. `from_raw_parts`) are still relevant for the call itself (that is: `from_raw_parts` would explode if we pass it a null pointer, so we need to check that first).

Reply:
Thank you very much for the explanation, much clearer now!